test/math_libraries/convolutions.py

```diff
@@ -0,0 +1,100 @@
+                weight_g = self._subtensor(weight, 0, group, g)
+                bias_g = self._subtensor(bias, 0, group, g)
+                outputs.append(self._thnn_conv(input_g, weight_g, k, bias_g, s, p, d))
+            return torch.cat(outputs, 1)
+
+    # Set multiple situations to improve convolution2d test coverage
+    @onlyOnCPUAndCUDA
+    @skipCUDAIfRocm
+    @dtypesIfCUDA(torch.half, torch.float, torch.double)
+    @dtypes(torch.float)
+    def test_conv2d_ext(self, device, dtype):
+        case_files = os.popen('ls shapes_*').read().strip().split()
+        total_cases = list()
+        for file_name in case_files:
+            with open(file_name) as f:
+                model_cases = json.loads(f.read())
+            total_cases += model_cases
+        for case in total_cases:
+            bs = case['mb']
+            group = case['g']
+            ic, ih, iw = case['ic'], case['ih'], case['iw']
+            oc = case['oc']
+            kh, kw = case['kh'], case['kw']
+            sh, sw = case['sh'], case['sw']
+            ph, pw = case['ph'], case['pw']
+            dh, dw = case['dh'], case['dw']
+            has_bias = case['bias']
+            if dh == 0 or dw == 0:
+                continue
+
+            ic_g = ic // group
+            torch.manual_seed(1)
+            input = torch.randn((bs, ic, ih, iw), device=device, dtype=dtype, requires_grad=True)
+            torch.manual_seed(2)
+            weight = torch.randn((oc, ic_g, kh, kw), device=device, dtype=dtype) * 0.01
+            torch.manual_seed(3)
+            bias = None if has_bias else torch.randn((oc), device=device, dtype=dtype)
+
+            k = [kh, kw]
+            s = [sh, sw]
+            p = [ph, pw]
+            d = [dh, dw]
+            thnn_output = self._thnn_conv_group(input, weight, k, bias, s, p, d, group)
+            if self.device_type == 'cpu' and torch.backends.mkldnn.is_available():
+                output = torch.mkldnn_convolution(input, weight, bias, p, s, d, group)
+            elif self.device_type == 'cuda' and torch.backends.cudnn.is_available():
+                output = torch.cudnn_convolution(input, weight, bias, p, s, d, group, True, True)
+            else:
+                output = torch.conv2d(input, weight, bias, s, p, d, group)
+
+            msg = 'device:{}, dtype:{}, group:{}, batchsize:{}' \
+                  'input channel:{}, output channel:{}, ' \
+                  'bias:{}, padding:{}, dilation:{}, stride:{}, ' \
+                  'kernel:{}'
+            msg = msg.format(device, dtype, group, bs, ic, oc, has_bias, p, d, s, k)
+
+            if self.device_type == 'cpu':
+                self.assertEqual(
+                    output, thnn_output,
+                    msg=msg
+                )
+            elif self.device_type == 'cuda':
+                self.assertAlmostEqual(
+                    output, thnn_output,
+                    msg=msg,
+                    delta=1e-2
```

Comment:
you can call assertEqual with corresponding arguments. This precision seems very high, if you need higher tolerance for half you can set it with @precisionOverride decorator